nixos/modules/services/networking/3proxy.nix

```diff
@@ -0,0 +1,460 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  pkg = pkgs._3proxy;
+  cfg = config.services._3proxy;
+  optionalList = list:
+    if list == [ ] then "*" else concatMapStringsSep "," toString list;
+  configFile = if cfg.confFile != null then
+    cfg.confFile
+  else
+    (pkgs.writeText "3proxy.conf" ''
+      # log to stdout
+      log
+
+      ${concatMapStringsSep "\n" (x: "nserver " + x) cfg.resolution.nserver}
+
+      nscache ${toString cfg.resolution.nscache}
+      nscache6 ${toString cfg.resolution.nscache6}
+
+      ${concatMapStringsSep "\n" (x: "nsrecord " + x)
+      (mapAttrsToList (name: value: "${name} ${value}")
+        cfg.resolution.nsrecord)}
+
+      ${optionalString (cfg.usersFile != null)
+        ''users $"${cfg.usersFile}"''
+      }
+      ${concatMapStringsSep "\n" (user:
+        ''
+          users "${
+            concatStringsSep ":" [ user.login user.passwordType user.password ]
+          }"'') cfg.users}
+
+      ${concatMapStringsSep "\n" (service: ''
+        auth ${concatStringsSep " " service.auth}
+
+        ${optionalString (cfg.denyPrivate)
+        "deny * * ${optionalList cfg.privateRanges}"}
+
+        ${concatMapStringsSep "\n" (acl:
+          "${acl.rule} ${
+            concatMapStringsSep " " optionalList [
+              acl.users
+              acl.sources
+              acl.targets
+              acl.targetPorts
+            ]
+          }") service.acl}
+
+        maxconn ${toString service.maxConnections}
+
+        ${optionalString (service.extraConfig != null) service.extraConfig}
+
+        ${service.type} -i${toString service.bindAddress} ${
+          optionalString (service.bindPort != null)
+          "-p${toString service.bindPort}"
+        } ${
+          optionalString (service.extraArguments != null) service.extraArguments
+        }
+
+        flush
+      '') cfg.services}
+      ${optionalString (cfg.extraConfig != null) cfg.extraConfig}
+    '');
+in with lib; {
+  options.services._3proxy = {
+    enable = mkEnableOption "3proxy";
+    confFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "/etc/3proxy.conf";
+      description = ''
+        Ignore all other 3proxy options and load configuration from this file.
+      '';
+    };
+    users = mkOption {
```

Comment:
I don't like options like this because it promotes bad user behavior and serves as more examples to other nixos developers on why it is ok to store secrets in the nix store (hint: it is not).

You have a `usersFile` option which covers the required functionality in a relatively simple way. What I'm suggesting is that you remove the `users` option in favor of `usersFile`. People who are entirely insistent on storing secrets in the nix store (or just want to quickly test with dummy data) are still free to write out something like:
```
services._3proxy.usersFile = pkgs.writeText ''
  user aanderse:CL:test
'';
```

I want to make it very clear this is just my opinion and not required in order to have this PR merged.

Reply:
Removed.